src/engine/mod.rs

```diff
@@ -0,0 +1,320 @@
+use self::context::Context;
+use self::equal::RelativeEq;
+use crate::{
+    builtin::{
+        filter::{self, FilterFn},
+        function::{self, FunctionFn},
+    },
+    error::{bail, Result},
+    parser::ast::*,
+    utils::validate_f64,
+};
+use serde_json::{Number, Value};
+use std::collections::HashMap;
+
+pub mod context;
+mod equal;
+
+pub struct Engine {
+    functions: HashMap<String, FunctionFn>,
+    filters: HashMap<String, FilterFn>,
+}
+
+impl Default for Engine {
+    fn default() -> Engine {
+        let mut filters = HashMap::new();
+        filters.insert("upper".into(), Box::new(filter::upper) as FilterFn);
+        filters.insert("lower".into(), Box::new(filter::lower) as FilterFn);
+
+        let mut functions = HashMap::new();
+        functions.insert("uuidv4".into(), Box::new(function::uuidv4) as FunctionFn);
+
+        Engine { functions, filters }
+    }
+}
+
+impl Engine {
+    fn eval_math(&self, lhs: &Number, rhs: &Number, operator: MathOperator) -> Result<Number> {
+        match operator {
+            MathOperator::Addition => {
+                if lhs.is_i64() && rhs.is_i64() {
+                    if let Some(x) = lhs.as_i64().unwrap().checked_add(rhs.as_i64().unwrap()) {
+                        return Ok(Number::from(x));
+                    }
+                }
+
+                let lhs = lhs.as_f64().unwrap();
+                let rhs = rhs.as_f64().unwrap();
+                let result = lhs + rhs;
+
+                Ok(Number::from_f64(validate_f64(result)?).unwrap())
+            }
+            MathOperator::Subtraction => {
+                if lhs.is_i64() && rhs.is_i64() {
```

Comment:
the patter of `if..as..unwrap...FUNCTION...as...return` seems to be used throughout this function, would it make sense to refactor to extract a function that takes a lambda ?

Reply:
Agree here, but will probably do it later if you don't mind. Did add TODO here.